RecoHI/HiEvtPlaneAlgos/src/EvtPlaneProducer.cc

```diff
@@ -314,107 +465,80 @@ EvtPlaneProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+	  // dz and d0 significance cuts 
+	  if ( fabs(dz/dzsigma) > 3 ) accepted = false;
+	  if ( fabs(d0/d0sigma) > 3 ) accepted = false;
+	  // pt resolution cut
+	  if ( j->ptError()/j->pt() > 0.1 ) accepted = false;
+	  // number of valid hits cut
+	  if ( j->numberOfValidHits() < 12 ) accepted = false;
+	}
 	if( accepted ) {
 	  track_eta = j->eta();
 	  track_phi = j->phi();
 	  track_pt = j->pt();
-	  double s =sin(2*track_phi);
-	  double c =cos(2*track_phi);
-	  double s3 =sin(3*track_phi);
-	  double c3 =cos(3*track_phi);
-	  double s4 =sin(4*track_phi);
-	  double c4 =cos(4*track_phi);
-	  double s5 =sin(5*track_phi);
-	  double c5 =cos(5*track_phi);
-	  double s6 =sin(6*track_phi);
-	  double c6 =cos(6*track_phi);
-	  double w = 1;
-	  if(useTrackPtWeight_) {
-	    w = track_pt;
-	    if(w>2.5) w=2.0;   //v2 starts decreasing above ~2.5 GeV/c
+	  double minpt = minpt_;
+	  double maxpt = maxpt_;
+	  for(int i = 0; i<NumEPNames; i++) {
+	    if(minpt_<0) minpt = minTransverse[i];
+	    if(maxpt_<0) maxpt = maxTransverse[i];
+	    if(track_pt<minpt) continue;
+	    if(track_pt>maxpt) continue;
+	    if(EPDet[i]==Tracker) {
+	      double w = track_pt;
+	      if(w>2.5) w=2.0;   //v2 starts decreasing above ~2.5 GeV/c
+	      if(EPOrder[i]==1) {
+		if(MomConsWeight[i][0]=='y' && loadDB_) {
+		  w = flat[i]->GetW(track_pt, vzr_sell, bin);
+		}
+		if(track_eta<0) w=-w;
+	      }
+	      rp[i]->addParticle(w,track_pt,sin(EPOrder[i]*track_phi),cos(EPOrder[i]*track_phi),track_eta);
+	    }
 	  }
-	  if(track_pt<minpt_) continue;
-	  if(track_pt>maxpt_) continue;
-	  rp[EvtPlaneFromTracksMidEta]->addParticle(w,s,c,track_eta);
-	  rp[EvtPTracksPosEtaGap]->addParticle(w,s,c,track_eta);
-	  rp[EvtPTracksNegEtaGap]->addParticle(w,s,c,track_eta);
-	  rp[EPTracksMid3]->addParticle(w,s3,c3,track_eta);
-	  rp[EPTracksPos3]->addParticle(w,s3,c3,track_eta);
-	  rp[EPTracksNeg3]->addParticle(w,s3,c3,track_eta);
-	  rp[EPTracksMid4]->addParticle(w,s4,c4,track_eta);
-	  rp[EPTracksPos4]->addParticle(w,s4,c4,track_eta);
-	  rp[EPTracksNeg4]->addParticle(w,s4,c4,track_eta);
-	  rp[EPTracksMid5]->addParticle(w,s5,c5,track_eta);
-	  rp[EPTracksPos5]->addParticle(w,s5,c5,track_eta);
-	  rp[EPTracksNeg5]->addParticle(w,s5,c5,track_eta);
-	  rp[EPTracksMid6]->addParticle(w,s6,c6,track_eta);
-	  rp[EPTracksPos6]->addParticle(w,s6,c6,track_eta);
-	  rp[EPTracksNeg6]->addParticle(w,s6,c6,track_eta);
-	  
-	}
-	
-      }
+	}  
+      } //end for
+    }
+    
+    std::auto_ptr<EvtPlaneCollection> evtplaneOutput(new EvtPlaneCollection);
+    EvtPlane *ep[NumEPNames];
+    
+    double ang=-10;
+    double sv = 0;
+    double cv = 0;
+    double svNoWgt = 0;
+    double cvNoWgt = 0;
+
+    double wv = 0;
+    double wv2 = 0;
+    double pe = 0;
+    double pe2 = 0;
+    uint epmult = 0;
+
+    for(int i = 0; i<NumEPNames; i++) {
+      rp[i]->getAngle(ang,sv,cv,svNoWgt, cvNoWgt, wv,wv2,pe,pe2,epmult);
+      ep[i] = new EvtPlane(i,0,ang,sv,cv,wv,wv2,pe,pe2,epmult);
```

Comment:
this is a memory leak.
Instead of making a temporary on the heap, just write directly to evtplaneOutput
```C++
evtplaneOutput->push_back(EvtPlane(i,0,ang,sv,cv,wv,wv2,pe,pe2,epmult));
evtplaneOutput->back().AddLevel(3, 0., svNoWgt, cvNoWgt);
```